File: diagrams/render.py
import os
from playwright.sync_api import sync_playwright

here = os.path.dirname(os.path.abspath(__file__))
url = "file://" + os.path.join(here, "funnel.html")
out = os.path.join(here, "funnel.png")

with sync_playwright() as p:
    browser = p.chromium.launch()
    ctx = browser.new_context(device_scale_factor=2, viewport={"width": 2400, "height": 700})
    page = ctx.new_page()
    page.goto(url)
    page.wait_for_timeout(300)
    el = page.query_selector(".wrap")
    el.screenshot(path=out)
    browser.close()
print("WROTE", out)

File: diagrams/render.js
const { chromium } = require('playwright-core');
const path = require('path');

(async () => {
  const url = 'file://' + path.join(__dirname, 'funnel.html');
  const out = path.join(__dirname, 'funnel.png');
  const browser = await chromium.launch({
    executablePath: '/opt/pw-browsers/chromium-1194/chrome-linux/chrome',
    args: ['--no-sandbox'],
  });
  const ctx = await browser.newContext({ deviceScaleFactor: 2, viewport: { width: 2400, height: 700 } });
  const page = await ctx.newPage();
  await page.goto(url, { waitUntil: 'networkidle' });
  await page.waitForTimeout(300);
  const el = await page.$('.wrap');
  await el.screenshot({ path: out });
  await browser.close();
  console.log('WROTE', out);
})();
